File: Scala/pe3.scala
// PE3 - find the largest prime factor of 600851475143
// Solution: 6857

import scala.math.BigInt._

object PE3 {
  def main(args: Array[String]) = {
    println("Answer: " + largestPrimeFactor(600851475143L,2))
  }

  def largestPrimeFactor(n: BigInt, x: BigInt): BigInt = {
    if (n == x) n
    else {
      if (n % x == 0) largestPrimeFactor(n/x, x)
      else largestPrimeFactor(n, x+1)
    }
  }

}

File: JavaScript/pe3.js
// PE3 - What is the largest prime factor of 600851475143?
// Solution: 6857

var largestPrimeFactor = function(n) {
  var i = 2
  var mid = Math.ceil(Math.sqrt(n))
  while (i < mid && i < n) {
    while (n % i == 0) {
      n /= i
    }
    i += 1
  }

  return n
}

console.log(largestPrimeFactor(600851475143))
